### ISSUE TITLE
/srv/fluffi/data/fluffiweb/app/static/bootstrap does not exist

### ISSUE THREAD

**Issue by maverickwoo** [siemens/fluffi#93]
After the commit https://github.com/siemens/fluffi/commit/cbc2cfdcfbd1663d9202cf37251c82942fb4ad25#diff-c65083bf5529e8821f2c7d386c933379, `/srv/fluffi/data/fluffiweb/app/static/bootstrap` does not exist in a fresh install. However, the generated html page still refers to it.

Examples quoted from `view-source:http://web.fluffi:8880/` after a fresh install:

1. The `head` block:
```html
  <head>

    <title>FLUFFI - Home</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <!-- Bootstrap -->
    <link href="/static/bootstrap/css/bootstrap.min.css?bootstrap=3.3.7.1.dev1" rel="stylesheet">

    <!--[if lt IE 9]>
          <script src="/static/html5shiv.min.js"></script>
          <script src="/static/respond.min.js"></script>
    <![endif]-->


<link href="/static/css/fluffi.css" rel="stylesheet" />
<link href="/static/images/favicon.ico" rel="shortcut icon">
<script src="/static/3rdParty/jquery/jquery-2.1.4.js"></script>
<script src="/static/3rdParty/socket.io-client/socket.io.slim.js"></script>


  </head>
```

2. After `footer` towards the end of `body`:
```html
<footer class="container">
    <p>&copy; Siemens AG, Corporate Technology 2020 | FLUFFI 1.0</p>
</footer>



    
<script src="/static/javascript/fuzzjobNavigation.js" type="text/javascript"></script>

    <script src="/static/bootstrap/jquery.min.js?bootstrap=3.3.7.1.dev1"></script>
    <script src="/static/bootstrap/js/bootstrap.min.js?bootstrap=3.3.7.1.dev1"></script>

  </body>
```

**Comment by TomSie** [comment]
Could you please have a look at this @jones718

**Comment by jones718** [comment]
Does it say "/srv/fluffi/data/fluffiweb/app/static/bootstrap does not exist" in the browsers console? 
And is the webgui running in a docker container?
I'm trying to reproduce the issue right now. Thanks!

**Comment by maverickwoo** [comment]
The browser console shows:
```
Loading failed for the <script> with source “http://web.fluffi:8880/static/bootstrap/jquery.min.js?bootstrap=3.3.7.1.dev1”. web.fluffi:8880:179:1
Loading failed for the <script> with source “http://web.fluffi:8880/static/bootstrap/js/bootstrap.min.js?bootstrap=3.3.7.1.dev1”. web.fluffi:8880:180:1
```

We run everything in a VM. I will post another comment that will help you repro on a fresh Ubuntu 18.04 VM.

**Comment by maverickwoo** [comment]
(Note: The instructions below are only for reproducing the bug. They are not sufficient to obtain a full Fluffi installation.)

With a fresh Ubuntu 18.04 VM:

```bash
# Personalize VM
cat >> ~/.ssh/authorized_keys # paste your key here
git config --global user.email {{your email}}
git config --global user.name {{your name}}

# Set up machine and account
sudo apt update
sudo apt install -y docker-compose docker.io git-lfs tig zip
sudo usermod -aG docker $(id -un)
git lfs install
sudo reboot

# Make /srv/fluffi
git clone --depth 1 -b master https://github.com/siemens/fluffi.git ./fluffigit
sudo mkdir -p /srv/fluffi
sudo chown -R $(id -u):$(id -g) /srv/fluffi
mkdir -p \
  /srv/fluffi/data/dnsmasq \
  /srv/fluffi/data/ftp/files/fluffi/linux/{x64,x86,arm32,arm64} \
  /srv/fluffi/data/ftp/files/fluffi/windows/{x64,x86} \
  /srv/fluffi/data/tftp
touch /srv/fluffi/data/dnsmasq/dnsmasq.leases
rsync -ai --delete --exclude={'data/ftp/files/archive','data/ftp/files/deploy','data/ftp/files/fluffi','data/ftp/files/initial/linux/activePackages','data/ftp/files/initial/linux/inactivePackages','data/ftp/files/initial/windows/activePackages','data/ftp/files/initial/windows/inactivePackages','data/ftp/files/initial/windows/ansible/Driver','data/ftp/files/initial/windows/ansible/VCRedistributables','data/ftp/files/odroid','data/ftp/files/SUT','data/ftp/files/tftp-roots','data/ftp/files/ubuntu-mirror','data/tftp','data/smb/files/server*','data/dnsmasq/dnsmasq.leases','data/mon/grafana','data/mon/influxdb'} ./fluffigit/srv/fluffi/ /srv/fluffi/

# Put /srv/fluffi under version control
cd /srv/fluffi
git init
git add .
git commit -m "Import after rsync"

# Add GlobalManager (assuming binaries is in ~/gm-bin.zip)
unzip ~/gm-bin.zip -d /srv/fluffi/data/fluffigm/
git commit -m "Add GlobalManager"

# Adjust permissions
chown -R $(id -u):$(id -g) /srv/fluffi/data/ftp/files
chown -R $(id -u):$(id -g) /srv/fluffi/data/smb/files
chown -R $(id -u):$(id -g) /srv/fluffi/data/tftp
find /srv/fluffi/data/smb/files -type d -print -exec chmod 777 {} \;
find /srv/fluffi/data/ftp/files -type d -print -exec chmod 777 {} \;
find /srv/fluffi/data/tftp -type d -print -exec chmod 777 {} \;
chmod 555 /srv/fluffi/data/ftp/files
chmod -R +x /srv/fluffi/data/fluffigm

# Run get_static_dependencies.sh
cd /srv/fluffi/data/fluffiweb/app
./get_static_dependencies.sh
cd /srv/fluffi
git add .
git commit -m "Run get_static_dependencies.sh"

# Replace IP addr of master
grep -lr 10.66.0.1 | xargs -n 1 sed -i s/10.66.0.1/$(ip route get 1.1.1.1 | gawk '/^1/ {print $7}')/g
git add .
git commit -m "Replace IP addr of master"

# Build containers while we still have Internet
docker-compose build

# Go offline
sudo systemctl stop systemd-resolved.service
sudo rm /etc/resolv.conf
echo "search fluffi
nameserver 127.0.0.1
" | sudo tee /etc/resolv.conf

# Bring up containers
docker-compose up --force-recreate 
```

In our experience, the first two times of running `docker-compose up --force-recreate` will loop at:
> polenext_1   | Polemarch not available retrying (2)...

The third time or so we will get:
> polenext_1   | Initialization Finished!

(Note: We will go investigate this looping later. Now coming back onto this issue.)

Finally, in our dev machine, we would execute `ssh -D 12345 {{VM}}` to set up a SOCKS proxy  to the VM on port 12345, then we will use a browser that has been configured to use this proxy to visit http://web.fluffi:8880/.

**Comment by jones718** [comment]
Okay a quick fix is to run the removed block from the get_dependencies script again so the bootstrap folder exists.
```
rm -rf bootstrap static/bootstrap
mkdir -p static/bootstrap
git clone https://github.com/twbs/bootstrap.git
cd bootstrap
git checkout 0b9c4a4007c44201dce9a6cc1a38407005c26c86
cd ..
cp -a bootstrap/dist/. static/bootstrap
rm -rf  bootstrap
```
The problem was that the webgui should use only bootstrap by the flask_bootstrap package.
Unfortunately, thanks to your issue we found out that this currently only works when running the webgui locally in dev mode but not inside the docker container(production mode/nginx). This problem did not occur on our site because the bootstrap folder was never deleted.

After installing bootstrap again there will still be a jquery.js missing but this will not affect the functionality of the webgui.

We will start investigate in this issue and try to fix it properly soon.

**Comment by p4cx** [comment]
> In our experience, the first two times of running `docker-compose up --force-recreate` will loop at:
> 
> > polenext_1   | Polemarch not available retrying (2)...
> 
> The third time or so we will get:
> 
> > polenext_1   | Initialization Finished!
> 
> (Note: We will go investigate this looping later. Now coming back onto this issue.)

This is by design, as it will take some time to launch Polemarch.

**Comment by maverickwoo** [comment]
> This is by design, as it will take some time to launch Polemarch.

@p4cx, we agree that Polemarch can take some time. The intriguing thing is we have observed two very different behaviors after we run `docker-compose up --force-recreate`. Sometimes we get "Initialization Finished!" after fewer than 10 retry messages, but some other times we need to wait 10+ minutes. In any case, at present we are quite OK with this behavior so far and we also cannot be sure whether it is due to our local (non-standard, as seen above) configuration or not.

In any case, we will come back to file another issue if we have a more useful error to report about this. The note above is merely meant to inform anyone who tries to reproduce the bootstrap bug that they can just re-run `docker-compose up --force-recreate` a few times and one of those times they will get "Initialization Finished!" very fast.

**Comment by maverickwoo** [comment]
> Okay a quick fix is to run the removed block from the get_dependencies script again so the bootstrap folder exists.

@jones718, yes, this is basically what we did with our local installation. After manually inserting the content of `bootstrap/dist` into `/srv/fluffi/data/fluffiweb/app/static/bootstrap` and also adding `/srv/fluffi/data/fluffiweb/app/static/bootstrap/jquery.min.js`, http://web.fluffi:8880/ appears to render correctly (to the extent possible with our current deployment...)

**Comment by p4cx** [comment]
>  but some other times we need to wait 10+ minutes 

This shouldn't be the case, but for debugging issue #96 I created a Polemarch Container and I didn't have to wait that long, max. 20 retry messages. But thanks for mention that odd behavior.

**Comment by TomSie** [comment]
Is this issue fixed by https://github.com/siemens/fluffi/pull/105 ?

**Comment by jones718** [comment]
Yes it is fixed by #105. 
I have added the missing bootstrap and jquery files to the dependency sript, because some components of the webui need it and can not use the flask_bootstrap package as expected.
